osc-server/src/main/java/org/osc/core/broker/service/validator/DeleteDistributedApplianceRequestValidator.java

```diff
@@ -53,13 +59,48 @@ public DistributedAppliance validateAndLoad(BaseDeleteRequest request) throws Ex
                             da.getId()));
         }
 
+        validateDAReferenceToSfcAndSecurityGroup(em, da);
+
         if (!da.getMarkedForDeletion() && request.isForceDelete()) {
             throw new VmidcBrokerValidationException(
                     "Distributed Appilance with ID "
                             + request.getId()
                             + " is not marked for deletion and force delete operation is applicable only for entries marked for deletion.");
         }
-
         return da;
     }
+
+    private void validateDAReferenceToSfcAndSecurityGroup(EntityManager em, DistributedAppliance da) throws Exception {
+        for (VirtualSystem vs : da.getVirtualSystems()) {
+            try {
+                validateVSReferenceToSfcAndSecurityGroup(em, vs);
+            } catch (VmidcBrokerValidationException ex) {
+                throw new VmidcBrokerValidationException(da.getName() + " " + ex.getMessage());
+            }
+        }
+    }
+
+    private void validateVSReferenceToSfcAndSecurityGroup(EntityManager em, VirtualSystem vs) throws Exception {
+        if (vs.getServiceFunctionChains() == null) {
+            return;
+        }
+        for (ServiceFunctionChain sfc : vs.getServiceFunctionChains()) {
+            validateSfcReferenceToSecurityGroup(em, sfc);
+        }
+    }
+
+    private void validateSfcReferenceToSecurityGroup(EntityManager em, ServiceFunctionChain sfc) throws Exception {
+        List<SecurityGroup> sgList = SecurityGroupEntityMgr.listSecurityGroupsBySfcId(em, sfc.getId());
+        String sgNames = sgList.stream().filter(sg -> !sg.getMarkedForDeletion()).map(sg -> sg.getName())
+                .collect(Collectors.joining(", "));
+        if (!sgNames.isEmpty()) {
+            throw new VmidcBrokerValidationException(String.format(
+                    "distributed appliance cannot be deleted which is currently referencing Service Function Chain '%s' and binded to a SecurityGroup(s) '%s'",
```

Comment:
I believe you missed my comment before... once again there is nothing in this method about DA. The message should not have any comment about DA. It should be something like "The Service Function Chain sfc_name is referenced by the Security Group sg_name" . Also avoid using `type` names in the messages: SecurityGroup -> security group, this is also consistent with your use of Service Function Chain (rather than SFC or ServiceFunctionChain).